### ISSUE TITLE
Primitives discussion

### ISSUE THREAD

**Issue by tkrajcar** [tkrajcar/carbonmu#8]
I'd like to get some thoughts from those who might be interested about the 'primitives' that should be supported out of the box - these are the data types that every game should have. For example, MUSH/MUX use 'Player', 'Thing', 'Room', and 'Exit' - with players and things being pretty damn similar. MUDs have items, NPCs, players, rooms, etc. Some MUD drivers also have 'blueprint items' which are interacted with slightly differently than regular items, and so on.

Obviously, as a fresh engine, we have the power to completely rethink this. I don't want to put more datatypes than we need to in CarbonMU core since part of the focus of this project is to build something modular - something like 'weapon' is too game-style-specific IMO - but I've also been building exclusively MUSH/MUX/MUSE games for 20 years, so I could use some input on what other core primitives we should support.

It feels to me like the traditional four (Player/Thing/Room/Exit) are a lock, but I'm open to disagreement, and also any new ideas. Just comment.

**Comment by tkrajcar** [comment]
Thanks to everyone who weighed in on this discussion. I'm back working on advancing CarbonMU again and am turning my attentions to data modeling.
After quite a bit of thought, I've come up with a couple decisions, some of which contradict previous information and some of which align well. Consider these an open RFC if you're interested in weighing in, but I'm feeling fairly good about these so far.

* I'm abandoning my previous goal to have a pluggable database back-end, and going to standardize on MongoDB. Yes, yes, I know, but it makes a lot more sense to me to be able to leverage all of the features of a proper database system without having to re-create them, like querying and indexing and so on. I'm not wild about having to introduce an external dependency, but I did want to always support MongoDB as the "robust option". 
* If you're wondering why Mongo and not something else, the main reasons are: a) I have prior experience modeling MUish data in both schemaed SQL and MongoDB, and have found a schemaless structure to be ideal, b) I've successfully used MongoDB for MU* data and had a pleasant experience, c)  installing and administrating MongoDB is quite straightforward.
* Conceptually, there will be two large categories of data: Tangible things and non. 
* Tangible things will be things that are represented in a game-world room, things players can see in rooms and interact with using 'core' commands - rooms, players, things and exits will be the primary classes here, although I'm sure there will be others, or more properly, behaviors that can be attached to them. They will share a certain amount of functionality, so still looking into the best way to implement that.
* Non-tangible things are things that are not in rooms, like channels, bboards, etc. There won't be much if any common functionality between independent classes, so the game's data model won't impose much if anything here.
